### PR TITLE
Adds implementation of FS-1062-Async-WithCancellation

### PR DESCRIPTION
**Pull request by TheAngryByrd**
reference design: https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1062-Async-WithCancellation.md

### CODE REVIEW

File: src/fsharp/FSharp.Core/async.fs [hunk 1]
```diff
@@ -1176,6 +1176,20 @@ namespace Microsoft.FSharp.Control
                                                      econt = (fun edi -> ctxt.cont (Choice2Of2 (edi.GetAssociatedSourceException()))))
                 computation.Invoke newCtxt)
 
+        static member WithCancellation(computation : Async<'T>, cancellationToken : CancellationToken) : Async<'T> = async {
+            let! ct2 = Async.CancellationToken
+            use cts = CancellationTokenSource.CreateLinkedTokenSource (cancellationToken, ct2)
+            let tcs = new TaskCompletionSource<'T>()
+            use _reg = cts.Token.Register (fun () -> tcs.TrySetCanceled() |> ignore)
+            let inner = async {
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/async.fs]: We generally use this style in the codebase:

```fsharp
let inner =
    async {
        // body
    }

Could you make this adjustment?
```

File: src/fsharp/FSharp.Core/async.fsi [hunk 2]
```diff
@@ -79,6 +79,12 @@ namespace Microsoft.FSharp.Control
         /// <returns>A computation that returns a choice of type T or exception.</returns>
         static member Catch : computation:Async<'T> -> Async<Choice<'T,exn>>
 
+        /// <summary>Creates a new async computation that allows cancellation with a provided CancellationToken.</summary>
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/async.fsi]: Nit: async ---> asynchronous

File: src/fsharp/FSharp.Core/async.fsi [hunk 3]
```diff
@@ -79,6 +79,12 @@ namespace Microsoft.FSharp.Control
...
+        ///
+        /// <param name="computation">The computation to run. </param>
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/async.fsi]: Nit: extra space

**Comment by cartermp** [comment]
@TheAngryByrd are you still interested in pursuing this work? We'd need the tests as I mentioned in the earlier comment.

**Comment by TheAngryByrd** [comment]
Oh jeez sorry this got away from me. I’ll take a stab tomorrow

**Comment by cartermp** [comment]
yeet

**Comment by cartermp** [comment]
Closing this out. @TheAngryByrd please ping if you'd like to continue with tests on this. And of course anyone else who might be interested, please feel free to take a stab.

**Pull request closed without merging**